File: packages/js/src/telemetry.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { TELEMETRY_DEFAULT_ENDPOINT } from "./constants.js";

export interface TelemetryOptions {
  enabled?: boolean;
  endpoint?: string;
  flushIntervalMs?: number;
  maxBatchSize?: number;
  fetch?: typeof fetch;
  clientIdPath?: string;
}

export type TelemetrySdkName = "js" | "mcp" | "python";
export type TelemetryEventType = "request" | "tool_call";

export interface TelemetryEvent {
  anonymous_client_id: string;
  sdk_name: TelemetrySdkName;
  sdk_version: string;
  event_type: TelemetryEventType;
  endpoint?: string;
  tool_name?: string;
  symbol_or_query?: string;
  http_status?: number;
  latency_ms: number;
  error_type?: string | null;
  timestamp_iso: string;
}

export type TelemetryRecordInput = Omit<
  TelemetryEvent,
  "anonymous_client_id" | "sdk_name" | "sdk_version" | "timestamp_iso"
>;

const DEFAULT_FLUSH_INTERVAL_MS = 5000;
const DEFAULT_MAX_BATCH = 20;
const POST_TIMEOUT_MS = 2000;

function envTrue(value: string | undefined): boolean {
  if (!value) return false;
  const v = value.toLowerCase();
  return v === "1" || v === "true" || v === "yes";
}

export function resolveEnabled(
  opt: boolean | undefined,
  env: NodeJS.ProcessEnv,
): boolean {
  if (envTrue(env.DO_NOT_TRACK)) return false;
  if (opt === true) return true;
  if (opt === false) return false;
  return envTrue(env.HARUSPEX_TELEMETRY);
}

function defaultClientIdPath(): string {
  const xdg = process.env.XDG_DATA_HOME;
  if (xdg) return join(xdg, "haruspex", "id");
  return join(homedir(), ".haruspex", "id");
}

function isUuidLike(s: string): boolean {
  return /^[0-9a-f-]{8,}$/i.test(s.trim()) && s.trim().length <= 64;
}

function makeUuid(): string {
  const c = (globalThis as { crypto?: { randomUUID?: () => string } }).crypto;
  if (c && typeof c.randomUUID === "function") return c.randomUUID();
  const bytes = new Uint8Array(16);
  for (let i = 0; i < 16; i++) bytes[i] = Math.floor(Math.random() * 256);
  bytes[6] = (bytes[6] & 0x0f) | 0x40;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;
  const hex = Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

export function loadOrCreateClientId(path?: string): string {
  const p = path ?? defaultClientIdPath();
  try {
    if (existsSync(p)) {
      const v = readFileSync(p, "utf-8").trim();
      if (isUuidLike(v)) return v;
    }
  } catch {}
  const id = makeUuid();
  try {
    mkdirSync(dirname(p), { recursive: true });
    writeFileSync(p, id, "utf-8");
  } catch {}
  return id;
}

export function templatize(method: string, path: string): string {
  const raw = path.split("?")[0];
  let p = raw;
  if (/^\/scores\/batch$/i.test(raw)) p = "/scores/batch";
  else if (/^\/scores\/[^/]+\/history$/i.test(raw)) p = "/scores/{symbol}/history";
  else if (/^\/scores\/[^/]+$/i.test(raw)) p = "/scores/{symbol}";
  else if (/^\/stocks\/[^/]+\/news$/i.test(raw)) p = "/stocks/{symbol}/news";
  else if (/^\/search$/i.test(raw)) p = "/search";
  return `${method.toUpperCase()} ${p}`;
}

export class TelemetryClient {
  enabled: boolean;
  private readonly apiKey: string;
  private readonly sdkName: TelemetrySdkName;
  private readonly sdkVersion: string;
  private readonly endpoint: string;
  private readonly flushIntervalMs: number;
  private readonly maxBatchSize: number;
  private readonly fetchImpl: typeof fetch;
  private readonly clientId: string;
  private queue: TelemetryEvent[] = [];
  private timer: ReturnType<typeof setInterval> | null = null;
  private shuttingDown = false;
  private inFlight: Promise<void> | null = null;

  constructor(opts: {
    apiKey: string;
    sdkName: TelemetrySdkName;
    sdkVersion: string;
    options?: TelemetryOptions;
  }) {
    const o = opts.options ?? {};
    this.enabled = resolveEnabled(o.enabled, process.env);
    this.apiKey = opts.apiKey;
    this.sdkName = opts.sdkName;
    this.sdkVersion = opts.sdkVersion;
    this.endpoint =
      o.endpoint ?? process.env.HARUSPEX_TELEMETRY_ENDPOINT ?? TELEMETRY_DEFAULT_ENDPOINT;
    this.flushIntervalMs = o.flushIntervalMs ?? DEFAULT_FLUSH_INTERVAL_MS;
    this.maxBatchSize = o.maxBatchSize ?? DEFAULT_MAX_BATCH;
    this.fetchImpl = o.fetch ?? globalThis.fetch;
    this.clientId = this.enabled ? loadOrCreateClientId(o.clientIdPath) : "";
    if (this.enabled) {
      this.startTimer();
    }
  }

  private startTimer(): void {
    if (this.timer) return;
    this.timer = setInterval(() => {
      void this.flush();
    }, this.flushIntervalMs);
    if (typeof this.timer === "object" && this.timer !== null && "unref" in this.timer) {
      (this.timer as { unref: () => void }).unref();
    }
  }

  record(ev: TelemetryRecordInput): void {
    if (!this.enabled || this.shuttingDown) return;
    const full: TelemetryEvent = {
      anonymous_client_id: this.clientId,
      sdk_name: this.sdkName,
      sdk_version: this.sdkVersion,
      timestamp_iso: new Date().toISOString(),
      ...ev,
    };
    this.queue.push(full);
    if (this.queue.length >= this.maxBatchSize) void this.flush();
  }

  async flush(): Promise<void> {
    if (this.queue.length === 0) return;
    if (this.inFlight) {
      try {
        await this.inFlight;
      } catch {}
    }
    const batch = this.queue.splice(0, this.queue.length);
    this.inFlight = this.send(batch).finally(() => {
      this.inFlight = null;
    });
    try {
      await this.inFlight;
    } catch {}
  }

  private async send(batch: TelemetryEvent[]): Promise<void> {
    if (typeof this.fetchImpl !== "function") return;
    const attempt = async (): Promise<Response | null> => {
      const ctrl = new AbortController();
      const t = setTimeout(() => ctrl.abort(), POST_TIMEOUT_MS);
      try {
        const r = await this.fetchImpl(this.endpoint, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ events: batch }),
          signal: ctrl.signal,
        });
        return r;
      } catch {
        return null;
      } finally {
        clearTimeout(t);
      }
    };
    try {
      let res = await attempt();
      if (!res || (res.status >= 500 && res.status < 600)) {
        res = await attempt();
      }
      try {
        if (res && res.body) {
          const reader = (res as unknown as { body: { cancel?: () => void } }).body;
          reader.cancel?.();
        }
      } catch {}
    } catch {}
  }

  async shutdown(): Promise<void> {
    if (this.shuttingDown) return;
    this.shuttingDown = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush();
    if (this.inFlight) {
      try {
        await this.inFlight;
      } catch {}
    }
  }
}

File: packages/python/src/haruspex/telemetry.py
"""Opt-in usage telemetry for the Haruspex Python SDK.

Disabled by default. Enable via env `HARUSPEX_TELEMETRY=1` or by passing
`telemetry=TelemetryOptions(enabled=True)` to the client. `DO_NOT_TRACK=1`
overrides everything.
"""

from __future__ import annotations

import atexit
import json
import os
import re
import threading
import time
import uuid
from collections.abc import Mapping
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Optional

import httpx

from haruspex.constants import TELEMETRY_DEFAULT_ENDPOINT

_DEFAULT_FLUSH_INTERVAL_S = 5.0
_DEFAULT_MAX_BATCH = 20
_POST_TIMEOUT_S = 2.0


@dataclass
class TelemetryOptions:
    enabled: Optional[bool] = None
    endpoint: Optional[str] = None
    flush_interval_s: float = _DEFAULT_FLUSH_INTERVAL_S
    max_batch_size: int = _DEFAULT_MAX_BATCH
    client_id_path: Optional[Path] = None
    http: Optional[httpx.Client] = None
    async_http: Optional[httpx.AsyncClient] = None


def _env_true(value: Optional[str]) -> bool:
    if not value:
        return False
    return value.strip().lower() in {"1", "true", "yes"}


def resolve_enabled(opt: Optional[bool], env: Mapping[str, str]) -> bool:
    if _env_true(env.get("DO_NOT_TRACK")):
        return False
    if opt is True:
        return True
    if opt is False:
        return False
    return _env_true(env.get("HARUSPEX_TELEMETRY"))


def _default_client_id_path() -> Path:
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / "haruspex" / "id"
    return Path.home() / ".haruspex" / "id"


_UUID_RE = re.compile(r"^[0-9a-fA-F-]{8,64}$")


def load_or_create_client_id(path: Optional[Path] = None) -> str:
    p = path or _default_client_id_path()
    try:
        if p.is_file():
            v = p.read_text(encoding="utf-8").strip()
            if _UUID_RE.match(v):
                return v
    except OSError:
        pass
    cid = str(uuid.uuid4())
    try:
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(cid, encoding="utf-8")
    except OSError:
        pass
    return cid


def templatize(method: str, path: str) -> str:
    raw = path.split("?", 1)[0]
    if re.fullmatch(r"/scores/batch", raw, re.IGNORECASE):
        templ = "/scores/batch"
    elif re.fullmatch(r"/scores/[^/]+/history", raw, re.IGNORECASE):
        templ = "/scores/{symbol}/history"
    elif re.fullmatch(r"/scores/[^/]+", raw, re.IGNORECASE):
        templ = "/scores/{symbol}"
    elif re.fullmatch(r"/stocks/[^/]+/news", raw, re.IGNORECASE):
        templ = "/stocks/{symbol}/news"
    elif re.fullmatch(r"/search", raw, re.IGNORECASE):
        templ = "/search"
    else:
        templ = raw
    return f"{method.upper()} {templ}"


def _make_event(
    *,
    client_id: str,
    sdk_name: str,
    sdk_version: str,
    event_type: str,
    endpoint: Optional[str] = None,
    tool_name: Optional[str] = None,
    symbol_or_query: Optional[str] = None,
    http_status: Optional[int] = None,
    latency_ms: float,
    error_type: Optional[str] = None,
) -> dict[str, Any]:
    ev: dict[str, Any] = {
        "anonymous_client_id": client_id,
        "sdk_name": sdk_name,
        "sdk_version": sdk_version,
        "event_type": event_type,
        "latency_ms": int(round(max(0.0, latency_ms))),
        "timestamp_iso": time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime()) + "Z",
        "error_type": error_type,
    }
    if endpoint is not None:
        ev["endpoint"] = endpoint
    if tool_name is not None:
        ev["tool_name"] = tool_name
    if symbol_or_query is not None:
        ev["symbol_or_query"] = symbol_or_query
    if http_status is not None:
        ev["http_status"] = http_status
    return ev


@dataclass
class _Common:
    api_key: str
    sdk_name: str
    sdk_version: str
    options: TelemetryOptions = field(default_factory=TelemetryOptions)


class TelemetryClient:
    """Sync telemetry client. Daemon thread flusher."""

    def __init__(
        self,
        *,
        api_key: str,
        sdk_name: str,
        sdk_version: str,
        options: Optional[TelemetryOptions] = None,
    ) -> None:
        opts = options or TelemetryOptions()
        self.enabled = resolve_enabled(opts.enabled, os.environ)
        self._api_key = api_key
        self._sdk_name = sdk_name
        self._sdk_version = sdk_version
        self._endpoint = (
            opts.endpoint
            or os.environ.get("HARUSPEX_TELEMETRY_ENDPOINT")
            or TELEMETRY_DEFAULT_ENDPOINT
        )
        self._flush_interval = opts.flush_interval_s
        self._max_batch = opts.max_batch_size
        self._http = opts.http
        self._owns_http = self._http is None
        self._client_id = (
            load_or_create_client_id(opts.client_id_path) if self.enabled else ""
        )
        self._queue: list[dict[str, Any]] = []
        self._lock = threading.Lock()
        self._wake = threading.Event()
        self._stop = threading.Event()
        self._thread: Optional[threading.Thread] = None
        if self.enabled:
            self._thread = threading.Thread(
                target=self._loop, name="haruspex-telemetry", daemon=True
            )
            self._thread.start()
            atexit.register(self.shutdown)

    def record(
        self,
        *,
        event_type: str,
        endpoint: Optional[str] = None,
        tool_name: Optional[str] = None,
        symbol_or_query: Optional[str] = None,
        http_status: Optional[int] = None,
        latency_ms: float,
        error_type: Optional[str] = None,
    ) -> None:
        if not self.enabled or self._stop.is_set():
            return
        ev = _make_event(
            client_id=self._client_id,
            sdk_name=self._sdk_name,
            sdk_version=self._sdk_version,
            event_type=event_type,
            endpoint=endpoint,
            tool_name=tool_name,
            symbol_or_query=symbol_or_query,
            http_status=http_status,
            latency_ms=latency_ms,
            error_type=error_type,
        )
        size_trigger = False
        with self._lock:
            self._queue.append(ev)
            if len(self._queue) >= self._max_batch:
                size_trigger = True
        if size_trigger:
            self._wake.set()

    def _drain(self) -> list[dict[str, Any]]:
        with self._lock:
            batch = self._queue
            self._queue = []
        return batch

    def _send(self, batch: list[dict[str, Any]]) -> None:
        if not batch:
            return
        http = self._http or httpx.Client(timeout=_POST_TIMEOUT_S)
        try:
            payload = json.dumps({"events": batch}).encode("utf-8")
            headers = {
                "Authorization": f"Bearer {self._api_key}",
                "Content-Type": "application/json",
            }
            for attempt in range(2):
                try:
                    resp = http.request(
                        "POST",
                        self._endpoint,
                        content=payload,
                        headers=headers,
                        timeout=_POST_TIMEOUT_S,
                    )
                    if 500 <= resp.status_code < 600 and attempt == 0:
                        continue
                    return
                except httpx.HTTPError:
                    if attempt == 0:
                        continue
                    return
        finally:
            if self._owns_http and http is not self._http:
                try:
                    http.close()
                except Exception:
                    pass

    def _loop(self) -> None:
        while not self._stop.is_set():
            self._wake.wait(self._flush_interval)
            self._wake.clear()
            batch = self._drain()
            if batch:
                try:
                    self._send(batch)
                except Exception:
                    pass

    def flush(self) -> None:
        batch = self._drain()
        if batch:
            try:
                self._send(batch)
            except Exception:
                pass

    def shutdown(self) -> None:
        if self._stop.is_set():
            return
        self._stop.set()
        self._wake.set()
        if self._thread is not None:
            self._thread.join(timeout=2.0)
        self.flush()


class AsyncTelemetryClient:
    """Async telemetry client. Schedules sends via asyncio."""

    def __init__(
        self,
        *,
        api_key: str,
        sdk_name: str,
        sdk_version: str,
        options: Optional[TelemetryOptions] = None,
    ) -> None:
        opts = options or TelemetryOptions()
        self.enabled = resolve_enabled(opts.enabled, os.environ)
        self._api_key = api_key
        self._sdk_name = sdk_name
        self._sdk_version = sdk_version
        self._endpoint = (
            opts.endpoint
            or os.environ.get("HARUSPEX_TELEMETRY_ENDPOINT")
            or TELEMETRY_DEFAULT_ENDPOINT
        )
        self._max_batch = opts.max_batch_size
        self._http = opts.async_http
        self._owns_http = self._http is None
        self._client_id = (
            load_or_create_client_id(opts.client_id_path) if self.enabled else ""
        )
        self._queue: list[dict[str, Any]] = []
        self._lock = threading.Lock()
        self._closed = False

    def record(
        self,
        *,
        event_type: str,
        endpoint: Optional[str] = None,
        tool_name: Optional[str] = None,
        symbol_or_query: Optional[str] = None,
        http_status: Optional[int] = None,
        latency_ms: float,
        error_type: Optional[str] = None,
    ) -> None:
        if not self.enabled or self._closed:
            return
        ev = _make_event(
            client_id=self._client_id,
            sdk_name=self._sdk_name,
            sdk_version=self._sdk_version,
            event_type=event_type,
            endpoint=endpoint,
            tool_name=tool_name,
            symbol_or_query=symbol_or_query,
            http_status=http_status,
            latency_ms=latency_ms,
            error_type=error_type,
        )
        flush_now = False
        with self._lock:
            self._queue.append(ev)
            if len(self._queue) >= self._max_batch:
                flush_now = True
        if flush_now:
            import asyncio

            try:
                loop = asyncio.get_running_loop()
                loop.create_task(self.flush())
            except RuntimeError:
                pass

    async def flush(self) -> None:
        with self._lock:
            batch = self._queue
            self._queue = []
        if not batch:
            return
        http = self._http or httpx.AsyncClient(timeout=_POST_TIMEOUT_S)
        try:
            payload = json.dumps({"events": batch}).encode("utf-8")
            headers = {
                "Authorization": f"Bearer {self._api_key}",
                "Content-Type": "application/json",
            }
            for attempt in range(2):
                try:
                    resp = await http.request(
                        "POST",
                        self._endpoint,
                        content=payload,
                        headers=headers,
                        timeout=_POST_TIMEOUT_S,
                    )
                    if 500 <= resp.status_code < 600 and attempt == 0:
                        continue
                    return
                except httpx.HTTPError:
                    if attempt == 0:
                        continue
                    return
        finally:
            if self._owns_http and http is not self._http:
                try:
                    await http.aclose()
                except Exception:
                    pass

    async def shutdown(self) -> None:
        if self._closed:
            return
        self._closed = True
        await self.flush()


__all__ = [
    "TelemetryOptions",
    "TelemetryClient",
    "AsyncTelemetryClient",
    "resolve_enabled",
    "load_or_create_client_id",
    "templatize",
]
